src/fsharp/FSharp.Core/async.fs

```diff
@@ -1250,31 +1250,28 @@ namespace Microsoft.FSharp.Control
                                 |> unfake)
                     | Some maxDegreeOfParallelism ->
                         let mutable i = -1
-                        let worker = MakeAsync (fun _ ->
-                            while i < tasks.Length do
+                        let rec worker (trampolineHolder : TrampolineHolder) =
+                            if i < tasks.Length then
```

Comment:
Why do you check for bounds twice here? Is it really needed?

Reply:
It saves an interlocked operation in the case that we're already at tasks.Length. It would probably be ok to just always do the interlocked operation and remove the first if check, just habit to write interlocks like this.

Reply:
Fair enough